File: Code/Native/Source/LightSystem/InternalLightManager.cpp

#include "InternalLightManager.h"

#include <algorithm>

NotifyCategoryDef(lightmgr, "");


InternalLightManager::InternalLightManager() {
    _cmd_list = NULL;
    _shadow_manager = NULL;
}

void InternalLightManager::add_light(PT(RPLight) light) {
    nassertv(_shadow_manager != NULL);

    if (light->has_slot()) {
        lightmgr_cat.error() << "InternalLightManager: cannot add light " 
                             << "since it already has a slot!" << endl;
        return;
    }

    // Reference the light because we store it
    light->ref();

    // Find a free slot
    size_t slot;
    if (!_lights.find_slot(slot)) {
        lightmgr_cat.error() << "All light slots used!" << endl;
    }

    // Reserve the slot
    light->assign_slot(slot);
    _lights.reserve_slot(slot, light);

    // Setup the shadows in case the light uses them
    if (light->get_casts_shadows()) {
        setup_shadows(light);
    }

    // Store the light on the gpu
    gpu_update_light(light);
}


void InternalLightManager::setup_shadows(RPLight* light) {
    light->init_shadow_sources();

    size_t num_sources = light->get_num_shadow_sources();
    size_t base_slot;
    if (!_shadow_sources.find_consecutive_slots(base_slot, num_sources)) {
        lightmgr_cat.error() << "Failed to find slot for shadow sources!" << endl;
        return;
    }

    for (int i = 0; i < num_sources; ++i) {
        ShadowSource* source = light->get_shadow_source(i);
        source->set_needs_update(true);

        // Assign the slot
        size_t slot = base_slot + i;
        _shadow_sources.reserve_slot(slot, source);
        source->set_slot(slot);
    }
}


void InternalLightManager::remove_light(PT(RPLight) light) {
    nassertv(_shadow_manager != NULL);
    
    if (!light->has_slot()) {
        lightmgr_cat.error() << "Cannot detach light, light has no slot!" << endl;
        return;
    }

    _lights.free_slot(light->get_slot());
    gpu_remove_light(light);
    light->remove_slot();

    if (light->get_casts_shadows()) {

        // Free the slots of all sources, and also unregister their regions from
        // the shadow atlas.
        for (size_t i = 0; i < light->get_num_shadow_sources(); ++i) {
            ShadowSource* source = light->get_shadow_source(i);
            if (source->has_slot()) {
                _shadow_sources.free_slot(source->get_slot());
            }
            if (source->has_region()) {
                _shadow_manager->get_atlas()->free_region(source->get_region());
            }
        }

        // Remove all sources of the light by emitting a consecutive remove command
        gpu_remove_consecutive_sources(light->get_shadow_source(0),
                                       light->get_num_shadow_sources());

        light->clear_shadow_sources();
    }

    // Since we referenced the light when we stored it, we
    // have to decrease the reference as well
    light->unref();
}

void InternalLightManager::gpu_remove_consecutive_sources(ShadowSource *first_source,
                                                          size_t num_sources) {
    GPUCommand cmd_remove(GPUCommand::CMD_remove_sources);
    cmd_remove.push_int(first_source->get_slot());
    cmd_remove.push_int(num_sources);
    _cmd_list->add_command(cmd_remove);
}

void InternalLightManager::gpu_remove_light(RPLight* light) {
    nassertv(_cmd_list != NULL);
    nassertv(light->has_slot());
    GPUCommand cmd_remove(GPUCommand::CMD_remove_light);
    cmd_remove.push_int(light->get_slot());
    _cmd_list->add_command(cmd_remove);
}

void InternalLightManager::gpu_update_light(RPLight* light) {
    nassertv(_cmd_list != NULL);
    nassertv(light->has_slot());
    GPUCommand cmd_update(GPUCommand::CMD_store_light);
    cmd_update.push_int(light->get_slot());
    light->write_to_command(cmd_update);
    light->unset_dirty_flag();
    _cmd_list->add_command(cmd_update);
}

void InternalLightManager::gpu_update_source(ShadowSource* source) {
    nassertv(_cmd_list != NULL);
    GPUCommand cmd_update(GPUCommand::CMD_store_source);
    cmd_update.push_int(source->get_slot());
    source->write_to_command(cmd_update);
    _cmd_list->add_command(cmd_update);
}

void InternalLightManager::update_lights() {
    for (auto iter = _lights.begin(); iter != _lights.end(); ++iter) {
        RPLight* light = *iter;
        if (light && light->is_dirty()) {
            light->update_shadow_sources();
            gpu_update_light(light);
        }
    }
}


void InternalLightManager::update_shadow_sources() {

    // Find all dirty shadow sources and make a list of them
    vector<ShadowSource*> _sources_to_update;

     for (auto iter = _shadow_sources.begin(); iter != _shadow_sources.end(); ++iter) {
        ShadowSource* source = *iter;
        if (source && source->get_needs_update()) {
            _sources_to_update.push_back(source);
        }
    }

    // Sort the sources based on their resolution, so that sources with a bigger
    // resolution come first. This helps to get a better packing on the shadow atlas.
    auto cmp_source = [](ShadowSource* a, ShadowSource* b){ return a->get_resolution() > b->get_resolution(); };
    std::sort(_sources_to_update.begin(), _sources_to_update.end(), cmp_source);

    // Get a handle to the atlas, will be frequently used
    ShadowAtlas *atlas = _shadow_manager->get_atlas();

    // Free the regions of all sources which will get updated. We have to take into
    // account that only a limited amount of sources can get updated per frame.
    size_t update_slots = min(_sources_to_update.size(),
                                   _shadow_manager->get_num_update_slots_left());
    for(size_t i = 0; i < update_slots; ++i) {
        if (_sources_to_update[i]->has_region()) {
           atlas->free_region(_sources_to_update[i]->get_region());
        }
    }

    // Find an atlas spot for all regions
    for (size_t i = 0; i < update_slots; ++i) {
        ShadowSource *source = _sources_to_update[i];

        if(!_shadow_manager->add_update(source)) {
            // In case the ShadowManager lied about the number of updates left
            lightmgr_cat.error() << "ShadowManager ensured update slot, but slot is taken!" << endl;
            break;
        }

        // We have an update slot, and are guaranteed to get updated as soon
        // as possible, so we can start getting a new atlas position.
        size_t region_size = atlas->get_required_tiles(source->get_resolution());
        LVecBase4i new_region = atlas->find_and_reserve_region(region_size, region_size);
        LVecBase4f new_uv_region = atlas->region_to_uv(new_region);
        source->set_region(new_region, new_uv_region);
        
        // Mark the source as updated
        source->set_needs_update(false);
        gpu_update_source(source);
    }
}

void InternalLightManager::update() {
    nassertv(_shadow_manager != NULL); // Not initialized yet!
    nassertv(_cmd_list != NULL);       // Not initialized yet!

    update_lights();
    update_shadow_sources();
}

File: rpcore/pynative/internal_light_manager.py
"""

RenderPipeline

Copyright (c) 2014-2016 tobspr <tobias.springer1@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

"""
from __future__ import print_function
from rplibs.six.moves import range  # pylint: disable=import-error

from panda3d.core import Vec3

from rpcore.pynative.pointer_slot_storage import PointerSlotStorage
from rpcore.pynative.gpu_command import GPUCommand

MAX_LIGHT_COUNT = 65535
MAX_SHADOW_SOURCES = 2048


class InternalLightManager(object):

    """ Please refer to the native C++ implementation for docstrings and comments.
    This is just the python implementation, which does not contain documentation! """

    def __init__(self):
        self._lights = PointerSlotStorage(MAX_LIGHT_COUNT)
        self._shadow_sources = PointerSlotStorage(MAX_SHADOW_SOURCES)
        self._cmd_list = None
        self._shadow_manager = None
        self._camera_pos = Vec3(0)
        self._shadow_update_distance = 100.0

    def get_max_light_index(self):
        return self._lights.get_max_index()

    max_light_index = property(get_max_light_index)

    def get_num_lights(self):
        return self._lights.get_num_entries()

    num_lights = property(get_num_lights)

    def get_num_shadow_sources(self):
        return self._shadow_sources.get_num_entries()

    num_shadow_sources = property(get_num_shadow_sources)

    def set_shadow_manager(self, shadow_manager):
        self._shadow_manager = shadow_manager

    def get_shadow_manager(self):
        return self._shadow_manager

    shadow_manager = property(get_shadow_manager, set_shadow_manager)

    def set_command_list(self, cmd_list):
        self._cmd_list = cmd_list

    def set_camera_pos(self, pos):
        self._camera_pos = pos

    def set_shadow_update_distance(self, dist):
        self._shadow_update_distance = dist

    def add_light(self, light):
        if light.has_slot():
            print("ERROR: Cannot add light since it already has a slot!")
            return

        slot = self._lights.find_slot()
        if slot < 0:
            print("ERROR: Could not find a free slot for a new light!")
            return

        light.assign_slot(slot)
        self._lights.reserve_slot(slot, light)

        if light.get_casts_shadows():
            self.setup_shadows(light)

        self.gpu_update_light(light)

    def setup_shadows(self, light):
        light.init_shadow_sources()
        light.update_shadow_sources()

        num_sources = light.get_num_shadow_sources()
        base_slot = self._shadow_sources.find_consecutive_slots(num_sources)
        if base_slot < 0:
            print("ERROR: Failed to find slot for shadow sources!")
            return

        for i in range(num_sources):
            source = light.get_shadow_source(i)
            source.set_needs_update(True)
            slot = base_slot + i
            self._shadow_sources.reserve_slot(slot, source)
            source.set_slot(slot)

    def remove_light(self, light):
        assert light is not None
        if not light.has_slot():
            print("ERROR: Could not detach light, light was not attached!")
            return

        self._lights.free_slot(light.get_slot())
        self.gpu_remove_light(light)
        light.remove_slot()

        if light.get_casts_shadows():

            for i in range(light.get_num_shadow_sources()):
                source = light.get_shadow_source(i)
                if source.has_slot():
                    self._shadow_sources.free_slot(source.get_slot())
                if source.has_region():
                    self._shadow_manager.get_atlas().free_region(source.get_region())
                    source.clear_region()

            self.gpu_remove_consecutive_sources(
                light.get_shadow_source(0), light.get_num_shadow_sources())

            light.clear_shadow_sources()

    def gpu_remove_consecutive_sources(self, first_source, num_sources):
        cmd_remove = GPUCommand(GPUCommand.CMD_remove_sources)
        cmd_remove.push_int(first_source.get_slot())
        cmd_remove.push_int(num_sources)
        self._cmd_list.add_command(cmd_remove)

    def gpu_remove_light(self, light):
        cmd_remove = GPUCommand(GPUCommand.CMD_remove_light)
        cmd_remove.push_int(light.get_slot())
        self._cmd_list.add_command(cmd_remove)

    def gpu_update_light(self, light):
        cmd_update = GPUCommand(GPUCommand.CMD_store_light)
        cmd_update.push_int(light.get_slot())
        light.write_to_command(cmd_update)
        light.set_needs_update(False)
        self._cmd_list.add_command(cmd_update)

    def gpu_update_source(self, source):
        cmd_update = GPUCommand(GPUCommand.CMD_store_source)
        cmd_update.push_int(source.get_slot())
        source.write_to_command(cmd_update)
        self._cmd_list.add_command(cmd_update)

    def update_lights(self):
        for light in self._lights.begin():
            if light.get_needs_update():
                if light.casts_shadows:
                    light.update_shadow_sources()
            self.gpu_update_light(light)

    def update_shadow_sources(self):
        sources_to_update = []

        for source in self._shadow_sources.begin():
            # if source and source.get_needs_update():
                # sources_to_update.append(source)
            if source:
                bounds = source.get_bounds()
                distance_to_camera = (self._camera_pos - bounds.get_center()) - bounds.get_radius()
                if distance_to_camera < self._shadow_update_distance:
                    sources_to_update.append(source)
                else:
                    if source.has_region():
                        self._shadow_manager.get_atlas().free_region(source.get_region())
                        source.clear_region()

        def get_source_score(source):
            dist = (source.get_bounds().get_center() - self._camera_pos).length()
            return -dist + (10**10 if source.has_region() else 0)

        sorted_sources = list(sorted(sources_to_update, key=get_source_score))

        atlas = self._shadow_manager.get_atlas()
        update_slots = min(
            len(sorted_sources),
            self._shadow_manager.get_num_update_slots_left())

        for i in range(update_slots):
            if sorted_sources[i].has_region():
                atlas.free_region(sorted_sources[i].get_region())

        for i in range(update_slots):
            source = sorted_sources[i]

            if not self._shadow_manager.add_update(source):
                print("ERROR: Shadow manager ensured update slot, but slot is taken!")
                break

            region_size = atlas.get_required_tiles(source.get_resolution())
            new_region = atlas.find_and_reserve_region(region_size, region_size)
            new_uv_region = atlas.region_to_uv(new_region)
            source.set_region(new_region, new_uv_region)
            source.set_needs_update(False)
            self.gpu_update_source(source)

    def update(self):
        self.update_lights()
        self.update_shadow_sources()
